src/clj/puppetlabs/puppetserver/shell_utils.clj

```diff
@@ -0,0 +1,96 @@
+  {:exit-code schema/Int
+   :stderr schema/Str
+   :stdout schema/Str})
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Internal
+
+(schema/defn ^:private ^:always-validate
+  validate-command
+  "Checks the command string to ensure that is is an absolute path, executable
+  and that the file exists. An exception is thrown is any of those are not the
+  case."
+  [command :- schema/Str]
+  (let [command-file (io/as-file command)]
+    (cond
+      (not (.isAbsolute command-file))
+      (throw (IllegalArgumentException.
+              (format "An absolute path is required, but %s is not an absolute path" command)))
+      (not (.exists command-file))
+      (throw (IllegalArgumentException.
+              (format "The referenced command '%s' does not exist" command)))
+      (not (.canExecute command-file))
+      (throw (IllegalArgumentException.
+              (format "The referenced command '%s' is not executable" command))))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Public
+
+(schema/defn ^:always-validate
+  execute-command :- ExecutionResult
+  "Execute the specified fully qualified command and any included options and
+  return the exit-code, and the contents of stdout and stderr for the command."
+  ([command :- schema/Str]
+   (execute-command command []))
+  ([command :- schema/Str
+    options :- [schema/Str]]
+   (validate-command command)
+   (try
+     (let [process (ProcessWrapper. (. (Runtime/getRuntime) exec (into-array (concat [command] options))))]
+       {:exit-code (.getExitCode process)
+        :stderr (.getErrorString process)
+        :stdout (.getOutputString process)})
+     (catch IOException e
+       (throw (IllegalStateException. (format "Exception while executing '%s': %s" command (.getMessage e))))))))
+
+(schema/defn ^:always-validate
+execute-command' :- ExecutionResult
+  "Execute the specified fully qualified command and any included options and
+  return the exit-code, and the contents of stdout and stderr for the command."
+  ([command :- schema/Str]
+   (execute-command' command [""]))
+  ([command :- schema/Str
+    options :- [schema/Str]]
+   (validate-command command)
+   (let [commandline (CommandLine. command)
+         executor (DefaultExecutor.)]
+     (with-open [
+                 errstream (ByteArrayOutputStream.)
+                 outstream (ByteArrayOutputStream.)]
+       (doto executor
+         (. setStreamHandler (PumpStreamHandler. outstream errstream))
+         (. setExitValues nil))
+       (let [exit-code
+             (. executor execute (. commandline addArguments (into-array options) false))]
+         {:exit-code exit-code
+          :stderr (slurp (ByteArrayInputStream. (.toByteArray errstream)))
+          :stdout (slurp (ByteArrayInputStream. (.toByteArray outstream)))})))))
+
+(schema/defn ^:always-validate
+execute-command'' :- ExecutionResult
```

Comment:
Aw had a discussion IRL with @haus - now I understand we're going to choose between one of these 3 implementations before merging